modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -2420,34 +2490,40 @@ private CompletableFuture<CompletableFuture<?>> applyUpdateCommand(
             );
 
             if (!cmd.full()) {
-                CompletableFuture<UUID> fut = applyCmdWithExceptionHandling(cmd).thenApply(res -> {
+                // TODO: https://issues.apache.org/jira/browse/IGNITE-20124 Temporary code below
+                synchronized (safeTime) {
+                    storageUpdateHandler.handleUpdate(
```

Comment:
Why is it that for non-full commands, we apply the command on the Primary first, and only then send the Raft command, but for full commands, the order is opposite?

Reply:
Because in case of full command we apply not the write intent but write itself. Writes doesn't require txn state resolution and thus visible to other transactions directly. All in all that means that we should prevent write visibility in case of replication failure. The reliable way to do it is to put replication prior to primary replica adjustment.

Why we need direct put on primary replica?
Not only because replication is finished when it's finished on majoirty that may miss primary but also because primary replica may not be a part of replication group.

Reply:
This explains why for 1PC we do 'send to RAFT first, write to the Primary storage second'. But why for !1PC do we do it in the opposite order?

Reply:
Replication is async anyway (it returns a future which we don't wait). For !1PC, before your fix it was (schematically):

```
var future = replicateCommand(cmd);
applyLocally(cmd);
return future;
```

Now it is

```
applyLocally(cmd);
var future = replicateCommand(cmd);
return future;
```
It seems that we did not wait for replication in the previous version of the code as well. But the order has changed. Why? :)

Reply:
Because in new version it's guaranteed that we will update storage and safeTime prior to replication. Before there was a race here.